cartographer_grpc/framework/rpc.cc

```diff
@@ -101,29 +128,95 @@ void Rpc::RequestStreamingReadIfNeeded() {
 }
 
 void Rpc::Write(std::unique_ptr<::google::protobuf::Message> message) {
-  response_ = std::move(message);
   switch (rpc_handler_info_.rpc_type) {
+    case ::grpc::internal::RpcMethod::BIDI_STREAMING:
+      // For BIDI_STREAMING enqueue the message into the send queue and
+      // start write operations if none are currently in-flight.
+      send_queue_.emplace(SendItem{std::move(message), ::grpc::Status::OK});
+      PerformWriteIfNeeded();
+      break;
+    case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
+      SendFinish(::grpc::Status::OK, std::move(message));
+      break;
+    case ::grpc::internal::RpcMethod::NORMAL_RPC:
+      SendFinish(::grpc::Status::OK, std::move(message));
+      break;
+    default:
+      LOG(FATAL) << "RPC type not implemented.";
+  }
+}
+
+void Rpc::SendFinish(::grpc::Status status,
+                     std::unique_ptr<::google::protobuf::Message> message) {
+  switch (rpc_handler_info_.rpc_type) {
+    case ::grpc::internal::RpcMethod::BIDI_STREAMING:
+      CHECK(!message);
+      finish_event_.pending = true;
+      server_async_reader_writer_->Finish(status, &finish_event_);
+      break;
     case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
-      server_async_reader_->Finish(*response_.get(), ::grpc::Status::OK,
-                                   &finish_event_);
       finish_event_.pending = true;
+      response_ = std::move(message);
+      SendUnaryFinish(server_async_reader_.get(), status, response_.get(),
+                      &finish_event_);
       break;
     case ::grpc::internal::RpcMethod::NORMAL_RPC:
-      server_async_response_writer_->Finish(*response_.get(),
-                                            ::grpc::Status::OK, &finish_event_);
       finish_event_.pending = true;
+      response_ = std::move(message);
+      SendUnaryFinish(server_async_response_writer_.get(), status,
+                      response_.get(), &finish_event_);
       break;
     default:
       LOG(FATAL) << "RPC type not implemented.";
   }
 }
 
+void Rpc::Finish(::grpc::Status status) {
+  switch (rpc_handler_info_.rpc_type) {
+    case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
+      SendFinish(status, nullptr /* message */);
+      break;
+    case ::grpc::internal::RpcMethod::NORMAL_RPC:
+      SendFinish(status, nullptr /* message */);
+      break;
+    case ::grpc::internal::RpcMethod::BIDI_STREAMING:
+      send_queue_.emplace(SendItem{nullptr, status});
+      PerformWriteIfNeeded();
+      break;
+    default:
+      LOG(FATAL) << "RPC type not implemented.";
+  }
+}
+
+void Rpc::PerformWriteIfNeeded() {
+  if (send_queue_.empty() || write_event_.pending) {
+    return;
```

Comment:
Since this item is skipped and generally we write only one item, can we be sure the queue is eventually written?

Reply:
As discussed offline, yes: we looked at the cases that may happen and the behavior seems to ensure that the queue is eventually emptied.